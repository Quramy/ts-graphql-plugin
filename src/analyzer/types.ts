import { TsGraphQLPluginConfigOptions } from '../types';
import { TypeGenAddonFactory } from '../typegen';

export type OperationType = 'query' | 'mutation' | 'subscription' | 'fragment' | 'complex' | 'other';

export type TsGraphQLPluginConfig = Omit<TsGraphQLPluginConfigOptions, 'typegen'> & {
  typegen: {
    addonFactories: TypeGenAddonFactory[];
  };
};

export interface ManifestDocumentEntry {
  fileName: string;
  type: OperationType;
  operationName?: string;
  fragmentName?: string;
  body: string;
  tag?: string;
  documentStart: { line: number; character: number };
  documentEnd: { line: number; character: number };
  templateLiteralNodeStart: { line: number; character: number };
  templateLiteralNodeEnd: { line: number; character: number };
}

export interface ManifestOutput {
  documents: ManifestDocumentEntry[];
}
