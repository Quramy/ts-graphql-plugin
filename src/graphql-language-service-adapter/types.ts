import ts from 'typescript';
import { GraphQLSchema } from 'graphql';
import { ScriptSourceHelper, ResolveResult } from '../ts-ast-util';
import { SchemaBuildErrorInfo } from '../schema-manager/schema-manager';

export type GetCompletionAtPosition = ts.LanguageService['getCompletionsAtPosition'];
export type GetSemanticDiagnostics = ts.LanguageService['getSemanticDiagnostics'];
export type GetQuickInfoAtPosition = ts.LanguageService['getQuickInfoAtPosition'];

export interface AnalysisContext {
  debug(msg: string): void;
  getScriptSourceHelper(): ScriptSourceHelper;
  getSchema(): GraphQLSchema | null | undefined;
  getSchemaOrSchemaErrors(): [GraphQLSchema, null] | [null, SchemaBuildErrorInfo[]];
  findTemplateNode(
    fileName: string,
    position: number,
  ): ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression | undefined;
  findTemplateNodes(fileName: string): (ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression)[];
  resolveTemplateInfo(fileName: string, node: ts.TemplateExpression | ts.NoSubstitutionTemplateLiteral): ResolveResult;
}
