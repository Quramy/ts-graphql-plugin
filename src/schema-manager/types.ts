export type SchemaConfig = {
  schema:
    | string
    | {
        file: {
          path: string;
        };
      }
    | {
        http: {
          url: string;
          method?: string;
          headers?: { [key: string]: string };
        };
      }
    | {
        http: {
          fromScript: string;
        };
      };
  localSchemaExtensions?: string[];
};

export interface SchemaManagerHost {
  getProjectRootPath(): string;

  getConfig(): SchemaConfig;

  fileExists(path: string): boolean;
  readFile(path: string, encoding?: string): string | undefined;
  watchFile(
    path: string,
    cb: (fileName: string) => void,
    interval: number,
  ): {
    close(): void;
  };

  log(msg: string): void;
}
