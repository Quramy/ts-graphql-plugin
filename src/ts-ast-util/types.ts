import ts from 'typescript';

/**
 *
 * Helper to write TypeScript source file.
 *
 */
export interface OutputSource {
  /**
   *
   * Absolute path of the output file.
   *
   */
  readonly outputFileName: string;

  /**
   *
   * Absolute path of directory of the output file.
   *
   */
  readonly outputDirName: string;

  /**
   *
   * Get all statements in this output source.
   *
   * @returns An array of statement AST node.
   *
   */
  readonly getStatements: () => ReadonlyArray<ts.Statement>;

  /**
   *
   * Add statement to this output source.
   *
   */
  readonly pushStatement: (statement: ts.Statement) => void;

  /**
   *
   * Replace a statement in this output source to specified new statement.
   *
   * @param statement - target statement node to replace.
   * @param newStatement - new statement.
   * @returns True when the statement of the first argument is found. Otherwise false.
   *
   */
  readonly replaceStatement: (statement: ts.Statement, newStatement: ts.Statement) => boolean;

  /**
   *
   * Remove a statement in this output source.
   *
   * @param statement - target statement node to remove.
   * @returns True when the statement is found. Otherwise false.
   *
   */
  readonly removeStatement: (statement: ts.Statement) => boolean;

  /**
   *
   * Add default import statement to this output source.
   *
   * @remarks
   * This method does nothing when the output source already has the import
   *
   * @example
   *
   * ```ts
   * outputSource.pushDefaultImportIfNeeded('MyType', './my-module');
   * console.log(ts.createPrinter().print(outputSource.toSourceFile())); // -> 'import MyType from "./my-module";'
   * ```
   *
   * @param identifierName - the identifier name to import.
   * @param from - the module name specifier to import from.
   * @returns True if the output source is mutated by this operation.
   *
   */
  readonly pushDefaultImportIfNeeded: (identifierName: string, from: string) => boolean;

  /**
   *
   * Add named import statement to the output source.
   *
   * @remarks
   * This method does nothing when the output source already has the same import statement.
   *
   * @example
   *
   * ```ts
   * outputSource.pushDefaultImportIfNeeded('MyType', './my-module');
   * console.log(ts.createPrinter().print(outputSource.toSourceFile())); // -> 'import { MyType } from "./my-module";'
   * ```
   *
   * @param identifierName - the identifier name to import.
   * @param from - the module name specifier to import from.
   * @returns True if the output source is mutated by this operation.
   *
   */
  readonly pushNamedImportIfNeeded: (identifierName: string, from: string) => boolean;

  /**
   *
   * Write comment to the top of this output source.
   *
   */
  readonly writeLeadingComment: (comment: string) => void;

  /**
   *
   * Create TypeScript SourceFile object from this output source.
   *
   */
  readonly toSourceFile: () => ts.SourceFile;
}

export type ComputePosition = (innerPosition: number) => {
  fileName: string;
  pos: number;
  isInOtherExpression?: boolean;
};

/**
 *
 * Serves the following information.
 *
 * - interpolated string
 * - position converting functions between TS source and GraphQL document
 *
 **/
export interface ResolvedTemplateInfo {
  combinedText: string;
  getInnerPosition: ComputePosition;
  getSourcePosition: ComputePosition;
  convertInnerPosition2InnerLocation: (pos: number) => { line: number; character: number };
  convertInnerLocation2InnerPosition: (location: { line: number; character: number }) => number;
}

export interface ResolveErrorInfo {
  fileName: string;
  start: number;
  end: number;
}

export interface ResolveResult {
  resolvedInfo?: ResolvedTemplateInfo;
  resolveErrors: ResolveErrorInfo[];
}

export interface ScriptSourceHelper {
  getAllNodes: (fileName: string, condition: (n: ts.Node) => boolean) => ts.Node[];
  getNode: (fileName: string, position: number) => ts.Node | undefined;
  getLineAndChar: (fileName: string, position: number) => ts.LineAndCharacter;
  resolveTemplateLiteral: (
    fileName: string,
    node: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression,
  ) => ResolveResult;
  updateTemplateLiteralInfo: (
    target: ResolvedTemplateInfo,
    range: { start: number; end: number },
    text?: string,
  ) => ResolvedTemplateInfo;
}
