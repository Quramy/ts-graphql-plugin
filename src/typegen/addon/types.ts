import ts from 'typescript';
import type {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLUnionType,
  GraphQLInterfaceType,
  GraphQLScalarType,
} from 'graphql/type';
import type { DocumentNode, OperationDefinitionNode, FragmentDefinitionNode } from 'graphql/language';
import { OutputSource } from '../../ts-ast-util/types';

type GraphQLFragmentTypeConditionNamedType = GraphQLObjectType | GraphQLUnionType | GraphQLInterfaceType;

type ToStrict<T> = { [P in keyof T]-?: T[P] };
export type StrictAddon = ToStrict<TypeGenVisitorAddon>;

/**
 *
 * Creates Type Generator addon object.
 * This factory function is called back for each GraphQL template string.
 *
 */
export interface TypeGenAddonFactory {
  /**
   *
   * @param context - @see TypeGenVisitorAddonContext
   * @returns Addon
   *
   */
  (context: TypeGenVisitorAddonContext): TypeGenVisitorAddon;
}

export interface TypeGenVisitorAddonContext {
  /**
   *
   * Referenced GraphQL schema object.
   *
   */
  readonly schema: GraphQLSchema;

  /**
   *
   * Utility object to mutate output source file. @see OutputSource
   *
   */
  readonly source: OutputSource;

  /**
   *
   * Where the GraphQL template string is located. @see ExtractedInfo
   *
   */
  readonly extractedInfo: ExtractedInfo;
}

/**
 *
 * Represents where the template string including the GraphQL operation or fragments are located.
 *
 */
export interface ExtractedInfo {
  /**
   *
   * File path of the .ts file which contains the GraphQL template string.
   *
   */
  readonly fileName: string;

  /**
   *
   * TypeScript source file object which contains the GraphQL template string.
   *
   */
  readonly tsSourceFile: ts.SourceFile;

  /**
   *
   * Template string AST node corresponding the GraphQL operation or fragments.
   *
   */
  readonly tsTemplateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
}

/**
 *
 * The addon object's methods are called back during the core type generator visits each GraphQL node in the template string.
 * And one addon object is generated for one template string.
 *
 * You don't have to implement all methods.
 *
 */
export interface TypeGenVisitorAddon {
  /**
   *
   * Reacts when the type generator visits GraphQL Scalar field.
   * This function can return corresponding TypeScript type node.
   *
   * @remarks
   * This function can not how to serialize/deserialize the custom scalar but how to map scalar to TypeScript type only.
   * You should consistence between runtime serialize behavior and type mapping.
   *
   * @example
   *
   * If your schema has a custom scalar type `Date` and your GraphQL client deserializes it to JavaScript native string,
   * you can implement this type mapping as the following:
   *
   * ```graphql
   * scalar Date
   * type Post {
   *   id: ID!
   *   publishedAt: Date!
   * }
   * ```
   *
   * ```ts
   * const factory: TypeGenAddonFactory = () => ({
   *   customScalar({ scalarType }) {
   *     if (scalarType.name === 'Date') {
   *       return ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword);
   *     }
   *   },
   * });
   * ```
   *
   * @param input - contains GraphQL Scalar type.
   * @return Corresponding TypeScript type node. If return undefined, the TypeScript type is determined by the core type generator.
   *
   */
  customScalar?: (input: CustomScalarInput) => CustomScalarOutput;

  /**
   *
   * Reacts when the type generator visits whole GraphQL document node.
   *
   * @param input - contains the GraphQL document AST node.
   *
   */
  document?: (input: DocumentInput) => void;

  /**
   *
   * Reacts when the type generator visits GraphQL operation definition node.
   * It's useful if you want to process query or variables types.
   *
   * @example
   *
   * ```ts
   * const factory: TypeGenAddonFactory = ({ source }) => ({
   *   operationDefinition({ tsResultNode }) {
   *     const additionalTypeStatement = createAwesomeType(tsResultNode);
   *     source.pushStatement(additionalTypeStatement);
   *   },
   * });
   * ```
   *
   * @param input - contains query/variable TypeScript type declaration node. @see OperationDefinionInput
   *
   */
  operationDefinition?: (input: OperationDefinionInput) => void;

  /**
   *
   * Reacts when the type generator visits GraphQL fragment definition node.
   *
   * @param input - contains fragment TypeScript type declaration node. @see FragmentDefinitionInput
   *
   */
  fragmentDefinition?: (input: FragmentDefinitionInput) => void;
}

/**
 *
 * Input object for `TypeGenVisitorAddon.customScalar` callback.
 *
 */
export interface CustomScalarInput {
  /**
   *
   * Definition of type scalar type.
   *
   */
  readonly scalarType: GraphQLScalarType;
}

/**
 *
 * Output object for `TypeGenVisitorAddon.customScalar` callback.
 *
 */
export type CustomScalarOutput = void | ts.TypeNode;

/**
 *
 * Input object for `TypeGenVisitorAddon.document` callback.
 *
 */
export interface DocumentInput {
  /**
   *
   * GraphQL document AST node.
   *
   */
  readonly graphqlNode: DocumentNode;
}

/**
 *
 * Input object for `TypeGenVisitorAddon.operationDefinition` callback.
 *
 */
export interface OperationDefinionInput {
  /**
   *
   * GraphQL AST node of this operation.
   *
   */
  readonly graqhqlNode: OperationDefinitionNode;

  /**
   *
   * Definition of the operation. One of Query or Mutation or Subscription
   *
   */
  readonly operationType: GraphQLObjectType;

  /**
   *
   * TypeScript AST node which represents the result type of this operation.
   *
   */
  readonly tsResultNode: ts.TypeAliasDeclaration;

  /**
   *
   * TypeScript AST node which represents the variable type of this operation.
   *
   */
  readonly tsVariableNode: ts.TypeAliasDeclaration;
}

/**
 *
 * Input object for `TypeGenVisitorAddon.fragmentDefinition` callback.
 *
 */
export interface FragmentDefinitionInput {
  /**
   *
   * GraphQL AST node of this fragment
   *
   */
  readonly graphqlNode: FragmentDefinitionNode;

  /**
   *
   * Named type(object or interface or union type) this fragment is specified on.
   *
   */
  readonly conditionType: GraphQLFragmentTypeConditionNamedType;

  /**
   *
   * TypeScript AST node which represents this fragment type.
   *
   */
  readonly tsNode: ts.TypeAliasDeclaration;
}
