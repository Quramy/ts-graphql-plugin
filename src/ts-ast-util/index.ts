import ts from 'typescript/lib/tsserverlibrary';

export * from './types';

export * from './template-expression-resolver';

export { createScriptSourceHelper } from './script-source-helper';

export type TagCondition = string;

export function findNode(sourceFile: ts.SourceFile, position: number): ts.Node | undefined {
  function find(node: ts.Node): ts.Node | undefined {
    if (position >= node.getStart() && position < node.getEnd()) {
      return ts.forEachChild(node, find) || node;
    }
  }
  return find(sourceFile);
}

export function findAllNodes(sourceFile: ts.SourceFile, cond: (n: ts.Node) => boolean): ts.Node[] {
  const result: ts.Node[] = [];
  function find(node: ts.Node) {
    if (cond(node)) {
      result.push(node);
      return;
    } else {
      ts.forEachChild(node, find);
    }
  }
  find(sourceFile);
  return result;
}

export function isTagged(node: ts.Node, condition: TagCondition) {
  if (!node || !node.parent) return false;
  if (!ts.isTaggedTemplateExpression(node.parent)) return false;
  const tagNode = node.parent;
  return tagNode.tag.getText() === condition;
}

export function isTemplateLiteralTypeNode(node: ts.Node): node is ts.TemplateLiteralTypeNode {
  // ts.isNoSubstitutionTemplateLiteral exists TypeScript >= 4.1
  return typeof ts.isTemplateLiteralTypeNode === 'function' && ts.isTemplateLiteralTypeNode(node);
}
