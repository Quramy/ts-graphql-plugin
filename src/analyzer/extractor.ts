import ts from 'typescript';
import { parse, print, DocumentNode, GraphQLError } from 'graphql';
import { visit } from 'graphql/language';
import { isTagged, ScriptSourceHelper, ResolvedTemplateInfo } from '../ts-ast-util';
import { ManifestOutput, ManifestDocumentEntry, OperationType } from './types';
import { ErrorWithLocation, ERROR_CODES } from '../errors';
import { detectDuplicatedFragments } from '../gql-ast-util';

export type ExtractorOptions = {
  removeDuplicatedFragments: boolean;
  scriptSourceHelper: ScriptSourceHelper;
  debug: (msg: string) => void;
};

export type ExtractTemplateResolveErrorResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  resolveTemplateError: {
    message: string;
    start: number;
    end: number;
  };
  documentNode: undefined;
  resolevedTemplateInfo: undefined;
  graphqlError: undefined;
};

export type ExtractGraphQLErrorResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  resolevedTemplateInfo: ResolvedTemplateInfo;
  resolveTemplateError: undefined;
  graphqlError: GraphQLError;
  resolveTemplateErrorMessage: undefined;
  documentNode: undefined;
};

export type ExtractSucceededResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  documentNode: DocumentNode;
  resolevedTemplateInfo: ResolvedTemplateInfo;
  resolveTemplateError: undefined;
  graphqlError: undefined;
  resolveTemplateErrorMessage: undefined;
};

export type ExtractResult = ExtractTemplateResolveErrorResult | ExtractGraphQLErrorResult | ExtractSucceededResult;

export class Extractor {
  private readonly _removeDuplicatedFragments: boolean;
  private readonly _helper: ScriptSourceHelper;
  private readonly _debug: (msg: string) => void;

  constructor({ debug, removeDuplicatedFragments, scriptSourceHelper }: ExtractorOptions) {
    this._removeDuplicatedFragments = removeDuplicatedFragments;
    this._helper = scriptSourceHelper;
    this._debug = debug;
  }

  extract(files: string[], tagName?: string): ExtractResult[] {
    const results: ExtractResult[] = [];
    this._debug('Extract template literals from: ');
    this._debug(files.map(f => ' ' + f).join(',\n'));
    files.forEach(fileName => {
      const nodes = this._helper
        .getAllNodes(fileName, node => ts.isTemplateExpression(node) || ts.isNoSubstitutionTemplateLiteral(node))
        .filter(node => (tagName ? isTagged(node, tagName) : true)) as (
        | ts.TemplateExpression
        | ts.NoSubstitutionTemplateLiteral
      )[];
      nodes.forEach(node => {
        const { resolvedInfo, resolveErrors } = this._helper.resolveTemplateLiteral(fileName, node);
        if (!resolvedInfo) {
          resolveErrors
            .filter(re => re.fileName === fileName)
            .forEach(resolveError => {
              results.push({
                fileName,
                templateNode: node,
                resolveTemplateError: {
                  message: ERROR_CODES.templateIsTooComplex.message,
                  start: resolveError.start,
                  end: resolveError.end,
                },
                resolevedTemplateInfo: undefined,
                graphqlError: undefined,
                documentNode: undefined,
              });
            });
        } else {
          results.push({
            fileName,
            templateNode: node,
            resolevedTemplateInfo: resolvedInfo,
          } as any);
        }
      });
    });
    return results.map(result => {
      if (!result.resolevedTemplateInfo) return result;
      try {
        const rawDocumentNode = parse(result.resolevedTemplateInfo.combinedText);
        if (!this._removeDuplicatedFragments) {
          return {
            ...result,
            documentNode: rawDocumentNode,
          } as ExtractSucceededResult;
        }
        const duplicatedInfo = detectDuplicatedFragments(rawDocumentNode);
        const updatedResolvedInfo = duplicatedInfo.reduce(
          (acc, fragmentInfo) => this._helper.updateTemplateLiteralInfo(acc, fragmentInfo),
          result.resolevedTemplateInfo,
        );
        const documentNode = parse(updatedResolvedInfo.combinedText);
        return {
          ...result,
          documentNode,
          resolevedTemplateInfo: updatedResolvedInfo,
        } as ExtractSucceededResult;
      } catch (error) {
        if (error instanceof GraphQLError) {
          return {
            ...result,
            graphqlError: error,
          } as ExtractGraphQLErrorResult;
        } else {
          throw error;
        }
      }
    });
  }

  pickupErrors(
    extractResults: ExtractResult[],
    { ignoreGraphQLError }: { ignoreGraphQLError: boolean } = { ignoreGraphQLError: false },
  ) {
    const errors: ErrorWithLocation[] = [];
    extractResults.forEach(r => {
      if (r.resolveTemplateError) {
        errors.push(
          new ErrorWithLocation(r.resolveTemplateError.message, {
            fileName: r.fileName,
            severity: 'Warn',
            content: r.templateNode.getSourceFile().getFullText(),
            start: r.resolveTemplateError.start,
            end: r.resolveTemplateError.end,
          }),
        );
      } else if (!ignoreGraphQLError && r.graphqlError) {
        const innerLoc = r.graphqlError.locations && r.graphqlError.locations[0];
        if (!innerLoc) {
          errors.push(
            new ErrorWithLocation(r.graphqlError.message, {
              fileName: r.fileName,
              content: r.templateNode.getSourceFile().getFullText(),
              start: r.templateNode.getStart(),
              end: r.templateNode.getEnd(),
            }),
          );
        } else if (r.resolevedTemplateInfo) {
          const info = r.resolevedTemplateInfo;
          const innerPosition = info.convertInnerLocation2InnerPosition({
            line: innerLoc.line - 1,
            character: innerLoc.column - 1,
          });
          const start = info.getSourcePosition(innerPosition).pos;
          errors.push(
            new ErrorWithLocation(r.graphqlError.message, {
              fileName: r.fileName,
              content: r.templateNode.getSourceFile().getFullText(),
              start,
              end: start + 1,
            }),
          );
        }
      }
    });
    return errors;
  }

  getDominantDefinition(result: ExtractSucceededResult) {
    let type: OperationType | undefined;
    const definedFragmentNames: string[] = [];
    const referencedFragmentNames: string[] = [];
    let operationName: string | undefined;
    visit(result.documentNode, {
      FragmentDefinition(node) {
        if (!type) {
          type = 'fragment';
        }
        definedFragmentNames.push(node.name.value);
      },
      FragmentSpread(node) {
        referencedFragmentNames.push(node.name.value);
      },
      OperationDefinition(node) {
        if (!type || type === 'fragment') {
          type = node.operation;
        } else {
          type = 'complex';
        }
        if (!operationName) {
          operationName = node.name ? node.name.value : 'ANONYMOUS_QUERY';
        } else {
          operationName = 'MULTIPLE_OPERATIONS';
        }
      },
    });
    const noReferedFragmentNames = definedFragmentNames.filter(defName =>
      referencedFragmentNames.every(n => defName !== n),
    );
    return { type, operationName, fragmentName: noReferedFragmentNames[noReferedFragmentNames.length - 1] };
  }

  toManifest(extractResults: ExtractResult[], tagName: string = ''): ManifestOutput {
    const documents = extractResults
      .filter(r => !!r.documentNode)
      .map(result => {
        const r = result as ExtractSucceededResult;
        const { type, operationName, fragmentName } = this.getDominantDefinition(r);
        return {
          fileName: r.fileName,
          type: type || 'other',
          operationName,
          fragmentName,
          body: print(r.documentNode!),
          tag: tagName,
          templateLiteralNodeStart: this._helper.getLineAndChar(r.fileName, r.templateNode.getStart()),
          templateLiteralNodeEnd: this._helper.getLineAndChar(r.fileName, r.templateNode.getEnd()),
          documentStart: this._helper.getLineAndChar(r.fileName, r.templateNode.getStart() + 1),
          documentEnd: this._helper.getLineAndChar(r.fileName, r.templateNode.getEnd() - 1),
        } as ManifestDocumentEntry;
      });
    return {
      documents,
    };
  }
}
