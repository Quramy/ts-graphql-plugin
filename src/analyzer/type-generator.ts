import path from 'path';
import ts from 'typescript';
import { GraphQLSchema } from 'graphql/type';

import { TsGqlError, ErrorWithLocation } from '../errors';
import { mergeAddons, TypeGenVisitor, TypeGenError, TypeGenAddonFactory, TypeGenVisitorAddonContext } from '../typegen';
import { dasherize } from '../string-util';
import { OutputSource, createOutputSource } from '../ts-ast-util';
import { Extractor, ExtractSucceededResult } from './extractor';

export type TypeGeneratorOptions = {
  prjRootPath: string;
  extractor: Extractor;
  debug: (msg: string) => void;
  tag: string | undefined;
  addonFactories: TypeGenAddonFactory[];
};

export class TypeGenerator {
  private readonly _prjRootPath: string;
  private readonly _extractor: Extractor;
  private readonly _tag: string | undefined;
  private readonly _addonFactories: TypeGenAddonFactory[];
  private readonly _debug: (msg: string) => void;
  private readonly _printer: ts.Printer;

  constructor({ prjRootPath, extractor, tag, addonFactories, debug }: TypeGeneratorOptions) {
    this._prjRootPath = prjRootPath;
    this._extractor = extractor;
    this._tag = tag;
    this._addonFactories = addonFactories;
    this._debug = debug;
    this._printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed, removeComments: false });
  }

  createAddon({
    schema,
    extractedResult,
    outputSource,
  }: {
    schema: GraphQLSchema;
    extractedResult: ExtractSucceededResult;
    outputSource: OutputSource;
  }) {
    const context: TypeGenVisitorAddonContext = {
      schema,
      source: outputSource,
      extractedInfo: {
        fileName: extractedResult.fileName,
        tsTemplateNode: extractedResult.templateNode,
        tsSourceFile: extractedResult.templateNode.getSourceFile(),
      },
    };
    const addons = this._addonFactories.map(factory => factory(context));
    return { addon: mergeAddons(addons), context };
  }

  generateTypes({ files, schema }: { files: string[]; schema: GraphQLSchema }) {
    const extractedResults = this._extractor.extract(files, this._tag);
    const extractedErrors = this._extractor.pickupErrors(extractedResults);
    if (extractedErrors.length) {
      this._debug(`Found ${extractedErrors.length} extraction errors.`);
    }
    const typegenErrors: TsGqlError[] = [];
    const visitor = new TypeGenVisitor({ schema });
    const outputSourceFiles: { fileName: string; content: string }[] = [];
    extractedResults.forEach(extractedResult => {
      if (extractedResult.documentNode) {
        const { type, fragmentName, operationName } = this._extractor.getDominantDefinition(extractedResult);
        if (type === 'complex') {
          const fileName = extractedResult.fileName;
          const content = extractedResult.templateNode.getSourceFile().getFullText();
          const start = extractedResult.templateNode.getStart();
          const end = extractedResult.templateNode.getEnd();
          const errorContent = { fileName, content, start, end };
          const error = new ErrorWithLocation('This document node has complex operations.', errorContent);
          typegenErrors.push(error);
          return;
        }
        const operationOrFragmentName = type === 'fragment' ? fragmentName : operationName;
        if (!operationOrFragmentName) return;
        const outputFileName = path.resolve(
          path.dirname(extractedResult.fileName),
          '__generated__',
          dasherize(operationOrFragmentName) + '.ts',
        );
        try {
          const outputSource = createOutputSource({ outputFileName });
          const { addon } = this.createAddon({ schema, outputSource, extractedResult });
          const outputSourceFile = visitor.visit(extractedResult.documentNode, { outputSource, addon });
          const content = this._printer.printFile(outputSourceFile);
          outputSourceFiles.push({ fileName: outputFileName, content });
          this._debug(
            `Create type source file '${path.relative(this._prjRootPath, outputFileName)}' from '${path.relative(
              this._prjRootPath,
              extractedResult.fileName,
            )}'.`,
          );
        } catch (error) {
          if (error instanceof TypeGenError) {
            const sourcePosition = extractedResult.resolevedTemplateInfo.getSourcePosition(error.node.loc!.start);
            if (sourcePosition.isInOtherExpression) return;
            const fileName = extractedResult.fileName;
            const content = extractedResult.templateNode.getSourceFile().getFullText();
            const start = sourcePosition.pos;
            const end = extractedResult.resolevedTemplateInfo.getSourcePosition(error.node.loc!.end).pos;
            const errorContent = { fileName, content, start, end };
            const translatedError = new ErrorWithLocation(error.message, errorContent);
            typegenErrors.push(translatedError);
          } else {
            throw error;
          }
        }
      }
    });
    return { errors: [...extractedErrors, ...typegenErrors], outputSourceFiles };
  }
}
