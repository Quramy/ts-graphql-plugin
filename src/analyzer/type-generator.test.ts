import { buildSchema } from 'graphql';
import { TypeGenerator } from './type-generator';
import { createTesintExtractor } from './testing/testing-extractor';
import { ExtractSucceededResult } from './extractor';
import { TypeGenAddonFactory } from '../typegen/addon/types';
import { createOutputSource } from '../ts-ast-util';

function createTestingTypeGenerator({
  files = [],
  addonFactories = [],
}: {
  files?: { fileName: string; content: string }[];
  addonFactories?: TypeGenAddonFactory[];
}) {
  const extractor = createTesintExtractor(files, true);
  const generator = new TypeGenerator({
    prjRootPath: '',
    tag: undefined,
    addonFactories,
    extractor,
    debug: () => {},
  });
  return { generator, extractor };
}

describe(TypeGenerator, () => {
  const schema = buildSchema(`type Query { hello: String! }`);
  describe(TypeGenerator.prototype.createAddon, () => {
    it('should create context for type-generator add on', () => {
      const { generator, extractor } = createTestingTypeGenerator({
        files: [
          {
            fileName: 'main.ts',
            content: 'export query = `query MyQuery { hello }`;',
          },
        ],
      });
      const result = extractor.extract(['main.ts']) as ExtractSucceededResult[];
      const { addon, context } = generator.createAddon({
        extractedResult: result[0],
        schema,
        outputSource: createOutputSource({ outputFileName: 'my-query.ts' }),
      });
      expect(context.extractedInfo.fileName).toBe('main.ts');
      expect(context.extractedInfo.tsSourceFile).toBeTruthy();
      expect(context.source.outputFileName).toBe('my-query.ts');
      expect(addon).toBeTruthy();
    });
  });

  describe(TypeGenerator.prototype.generateTypes, () => {
    it('should create type files', () => {
      const { generator } = createTestingTypeGenerator({
        files: [
          {
            fileName: 'main.ts',
            content: 'const query = gql`query MyQuery { hello }`;',
          },
        ],
      });
      const { outputSourceFiles } = generator.generateTypes({
        files: ['main.ts'],
        schema,
      });
      if (!outputSourceFiles) return fail();
      expect(outputSourceFiles.length).toBe(1);
      expect(outputSourceFiles[0].fileName.endsWith('__generated__/my-query.ts')).toBeTruthy();
      expect(outputSourceFiles[0].content).toMatchSnapshot();
    });

    it('should ignore complex operations document', () => {
      const { generator } = createTestingTypeGenerator({
        files: [
          {
            fileName: 'main.ts',
            content: `
              const query = gql\`
                query MyQuery {
                  hello
                }
                mutation MyMutaion {
                  bye
                }
              \`;
            `,
          },
        ],
      });
      const { outputSourceFiles, errors } = generator.generateTypes({
        files: ['main.ts'],
        schema,
      });
      if (!outputSourceFiles) return fail();
      expect(outputSourceFiles.length).toBe(0);
      expect(errors.length).toBe(1);
      expect(errors[0].message).toMatchSnapshot();
    });

    it('should report errors occuring in typegen visitor', async () => {
      const { generator } = createTestingTypeGenerator({
        files: [
          {
            fileName: 'main.ts',
            content: 'const query = gql`query MyQuery { goodBye }`;',
          },
        ],
      });
      const { outputSourceFiles, errors } = generator.generateTypes({
        files: ['main.ts'],
        schema,
      });
      if (!outputSourceFiles) return fail();
      expect(outputSourceFiles.length).toBe(0);
      expect(errors.length).toBe(1);
      expect(errors[0].message).toMatchSnapshot();
    });
  });
});
