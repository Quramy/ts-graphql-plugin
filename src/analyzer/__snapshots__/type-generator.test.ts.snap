// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeGenerator generateTypes should create type files 1`] = `
"/* eslint-disable */
/* This is an autogenerated file. Do not edit this file directly! */
export type MyQuery = {
    hello: string;
};
export type MyQueryVariables = {};
"
`;

exports[`TypeGenerator generateTypes should ignore complex operations document 1`] = `"This document node has complex operations."`;

exports[`TypeGenerator generateTypes should report errors occuring in typegen visitor 1`] = `"Type \\"Query\\" does not have field \\"goodBye\\"."`;
